SalesforceCommerceCloud/pwa-kit
# Disable local development proxy when running remotely

To support local development, we provide a "mock" of Managed Runtime's proxy feature in the Express server.

In Managed Runtime, when configured correctly, proxies are *not* handled by the Express server and instead are routed through the CDN.

Currently, we always run the "mock" proxy, even in the context of Managed Runtime. 

When folks forget to configure their proxies in Managed Runtime (which is easier to do!), this leads to subtle bugs:

* duplicate query string parameters are not respected
* it is slow

This issue may not be clear until after a project has launched.

This change adds a new parameter `enableLegacyRemoteProxying` to `createApp` whose value defaults to `true`.

When set to `false`, local development proxies are _not_  added when the code is running in Managed Runtime. Folks that have not configured their proxies get a loud error and navigating directly to the proxy endpoints make it clear they are not configured correctly.

## For Discussion

Ideally, we be able to show a clear error message to developers in the context of the 500 page that will be returned from the SSR server, but as the stack trace is not available when running remotely, it isn't immediately obvious how to do that. It would be good to discuss how we'd like to propagate this sort of information to users.

## Types of Changes


## Changes

- Adds a new parameter `createApp({enableLegacyRemoteProxying})` that defaults to `true`. _New projects created with `pwa-kit-create-app` have the value set to `false` in `ssr.js`_.

## How to Test-Drive This PR

- Run the code locally, check that the PWA Kit still works.
- Review these two test environments:
- https://demo-proxies-enabled.mobify-storefront.com/ - has proxies set. Observe it still works.
- https://demo-proxies-disabled.mobify-storefront.com/ - does NOT have proxy set. Observe it is broken.
- https://demo-proxies-disabled.mobify-storefront.com/mobify/proxy/api/hello - Observe the custom error for sites without proxies setup correctly.

## General
